### PR TITLE
Adds prefix and suffix support to TextField, per Material Design spec.

### LINKED ISSUE #10559: Add prefix/suffix support to TextField
Per the Material design spec:

https://material.io/guidelines/components/text-fields.html#text-fields-input-types

### PR DESCRIPTION
**Pull request by gspencergoog**
This PR implements the prefix and suffix portions of the [Material Design spec for TextField](https://material.io/guidelines/components/text-fields.html#text-fields-input-types). It also fixes #10559.

The new interface is on the InputDecoration, adding four fields: prefixText, suffixText, prefixStyle, suffixStyle, which supply the text and text style for each.

If there is a hint or label in the way (e.g. it's supposed to be drawn in the text field), then the prefix/suffix text won't appear, but otherwise it will.  So, for instance, if you have just hint text, then the prefix/suffix won't appear until you enter something in the field, since the hint needs to be displayed until then.  If you have no hint text, but a label, then the prefix/suffix will appear whenever the label isn't inside the field (i.e. whenever it's focused).  If you have neither, then the prefix/suffix is there all the time.

I also updated the gallery and added tests.

### CODE REVIEW

File: packages/flutter/lib/src/material/input_decorator.dart [hunk 1]
```diff
@@ -37,6 +37,10 @@ class InputDecoration {
     this.errorStyle,
     this.isDense: false,
     this.hideDivider: false,
+    this.prefixText,
+    this.prefixStyle,
+    this.suffixText,
+    this.suffixStyle
```
Comment (Hixie) [inline packages/flutter/lib/src/material/input_decorator.dart]: nit: trailing comma

File: packages/flutter/lib/src/material/input_decorator.dart [hunk 2]
```diff
@@ -133,6 +141,28 @@ class InputDecoration {
   /// Defaults to false.
   final bool hideDivider;
 
+  /// Optional text prefix to place on the line before the input.
+  ///
+  /// Uses the [prefixStyle].  Uses [hintStyle] if [prefixStyle] isn't
```
Comment (Hixie) [inline packages/flutter/lib/src/material/input_decorator.dart]: nitty nit nit: here and below, only one space after a period

File: packages/flutter/lib/src/material/input_decorator.dart [hunk 3]
```diff
@@ -147,6 +177,10 @@ class InputDecoration {
     TextStyle errorStyle,
     bool isDense,
     bool hideDivider,
+    String prefixText,
+    TextStyle prefixStyle,
+    String suffixText,
+    TextStyle suffixStyle
```
Comment (Hixie) [inline packages/flutter/lib/src/material/input_decorator.dart]: nit: trailing comma

File: packages/flutter/lib/src/material/input_decorator.dart [hunk 4]
```diff
@@ -402,11 +456,29 @@ class InputDecorator extends StatelessWidget {
       );
     }
 
+    Widget inputChild;
+    if (!hasInlineLabel && (!isEmpty || hintText == null) &&
+        (decoration?.prefixText != null || decoration?.suffixText != null)) {
+      final List<Widget> rowContents = <Widget>[];
+      if (decoration.prefixText != null) {
+        rowContents.add(new Text(decoration.prefixText,
```
Comment (Hixie) [inline packages/flutter/lib/src/material/input_decorator.dart]: nit: line break before the `new` and before the matching `)`

File: packages/flutter/lib/src/material/input_decorator.dart [hunk 5]
```diff
@@ -402,11 +456,29 @@ class InputDecorator extends StatelessWidget {
...
+            style: decoration.prefixStyle ?? hintStyle));
+      }
+      rowContents.add(new Expanded(child: child));
+      if (decoration.suffixText != null) {
+        rowContents.add(new Text(decoration.suffixText,
```
Comment (Hixie) [inline packages/flutter/lib/src/material/input_decorator.dart]: ditto

File: packages/flutter/test/material/text_field_test.dart [hunk 6]
```diff
@@ -830,7 +1072,7 @@ void main() {
     newPos = tester.getTopLeft(find.text('Second'));
     expect(newPos.dy, lessThan(pos.dy));
   });
-
+
```
Comment (Hixie) [inline packages/flutter/test/material/text_field_test.dart]: nit: remove trailing spaces

**Comment by Hixie** [comment]
I'd rather we didn't have the changes to the `.iml` files (in fact I'd rather we didn't have them at all, but that's a problem for another PR!).

This patch looks great! Thanks so much!

File: packages/flutter/lib/src/material/input_decorator.dart (hunk 2 above)
Comment (gspencergoog) [inline packages/flutter/lib/src/material/input_decorator.dart]: LOL!  OK.  I blame my fingers.

**Review by gspencergoog — COMMENTED** [review]
OK, I think I took care of everything.  Thanks for the quick review!

File: packages/flutter/lib/src/material/input_decorator.dart (hunk 4 above)
Comment (Hixie) [inline packages/flutter/lib/src/material/input_decorator.dart]: dartfmt doesn't implement Flutter style so we can't use it. We are looking into other options but don't currently have one.
https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo describes Flutter style.

**Comment by Hixie** [comment]
LGTM

**Comment by Hixie** [comment]
Let me know if you'd like me to land this for you.

**Comment by Hixie** [comment]
Landed, thanks!

**Comment by Hixie** [comment]
This slightly regressed the flutter_gallery__memory_nav diff_total_kb and flutter_gallery__memory_nav end_total_kb benchmarks, but presumably that's because we're now allocating slightly more to run the relevant gallery tests. We went from around 135,023KB to around 137,148KB.

**Comment by eseidelGoogle** [comment]
How would this small change cause 2MB of memory usage regression?  That seems odd.

**Comment by Hixie** [comment]
We have a 10MB variance on the complex_layout_scroll_perf__memory start_total_kb benchmark from run to run, so 2MB is not huge. Maybe this increased our memory usage by 1KB per frame and that pushed us into the next GC pool size or something.

**Pull request merged**